models/metrics.py

```
@@ -90,6 +92,63 @@ def filter_by_time_of_day(df: pd.DataFrame, start_time_str, end_time_str) -> pd.
+                "closest_first_after_headway": timetable[timetable["arrival_time"] == s.first_scheduled_after_arrival].arrival_headway.values[0]
+            })
+
+        stops_df["closest_scheduled_arrival"] = stops_df["TIME"].apply(get_closest_timetable_time)
+        stops_df["first_scheduled_after_arrival"] = stops_df["TIME"].apply(get_closest_nonnegative_timetable_time)
+
+        stops_df["closest_delta"] = stops_df["TIME"] - stops_df["closest_scheduled_arrival"]
+        stops_df["first_after_delta"] = stops_df["TIME"] - stops_df["first_scheduled_after_arrival"]
+        stops_df[["closest_delta", "first_after_delta"]] = stops_df[["closest_delta", "first_after_delta"]].applymap(lambda x: x.total_seconds()/60 if isinstance(x, timedelta) else np.nan)
+
+        stops_df[["closest_scheduled_headway", "closest_first_after_headway"]] = stops_df.apply(get_corresponding_scheduled_headway, axis = "columns")
+        stops_df[["closest_scheduled_headway", "closest_first_after_headway"]] = stops_df[["closest_scheduled_headway", "closest_first_after_headway"]].applymap(lambda x: x.total_seconds()/60 if isinstance(x, timedelta) else np.nan)
+        
+        stops_df["closest_headway_delta"] = stops_df["headway"] - stops_df["closest_scheduled_headway"]
+        stops_df["first_headway_delta"] = stops_df["headway"] - stops_df["closest_first_after_headway"]
+
+        stops_df = stops_df.rename(mapper = {"TIME": "actual_arrival_time"}, axis = "columns")
+        
+        return stops_df[["actual_arrival_time", "closest_scheduled_arrival", "closest_delta", "first_scheduled_after_arrival", "first_after_delta", "headway", "closest_scheduled_headway", "closest_headway_delta",  "closest_first_after_headway", "first_headway_delta"]]
+    except Exception as err:
+        print(f"Error occurred while comparing timetable and actual data: {repr(err)}")
+        return pd.DataFrame()

```

Swallowing exceptions makes debugging difficult, could it just let the exception propagate to the caller?